### DISCUSSION TITLE
A few questions about this Pallas mamba kernel

### DISCUSSION THREAD

**Discussion by davisyoshida** [jax-ml/jax#19353] [Q&A]
(@sharadmv)

I made an attempt at implementing the Mamba block forward pass in Pallas [here](https://github.com/srush/annotated-mamba/issues/1#issuecomment-1890381520), but as you can see I had to drop down into Triton a bunch of times. A few questions:

1. I wrote some custom lowering rules for `bitcast_convert_type` and `convert_element_type` (to handle uint values), would you guys like a PR with support for these?
2. Not necessary for this version of the kernel, but is there any chance of getting strided loads supported? When I was trying to implement slicing I had to write custom lowering rules to do the pointer math for `tl.load`.
3. Any general advice on ways this could be made less hacky?

**Comment by sharadmv** [comment]
Thanks a lot for pushing on this Davis! I'm super excited about Mamba and your progress.

> I wrote some custom lowering rules for bitcast_convert_type and convert_element_type (to handle uint values), would you guys like a PR with support for these?

Of course, that would be amazing! Just ask that you write a couple tests in `pallas_test.py`.

> Not necessary for this version of the kernel, but is there any chance of getting strided loads supported? When I was trying to implement slicing I had to write custom lowering rules to do the pointer math for tl.load.

Yea we should support this. Let's perhaps open a new thread about the best way to surface these types of loads.

> Any general advice on ways this could be made less hacky?

Re: the `mamba_scan` primitive. Is is possible to repurpose `jax.lax.scan` for this? @mattjj. Not sure if we need a proper `associative_scan` primitive for this (even if it's just used in Pallas). Also, re AD. I think you're best off writing a backwards pass pallas kernel for this. I think ADing the pallas kernel is going to be tricky (currently only fwd mode works and only pretty simple kernels will AD efficiently once transposes are supported).

**Reply by davisyoshida** [reply]
I've been trying to get the tests running, but I'm running into the same issues with finding a version set that works. Do you know what version of triton/jax-triton is expected by JAX at head right now?

**Reply by sharadmv** [reply]
I would have to find a compatible Triton but I know recently the Triton version was updated. Maybe try one from ~1 week ago?

**Reply by davisyoshida** [reply]
Any guidance on how to get a version of jaxlib with the `jaxlib.triton.compat` package? The nightly jaxlib wheels don't include it, and when I built a wheel from source it didn't either.

**Reply by sharadmv** [reply]
I think this is an issue with the nightly builds that will be fixed soon. Sorry about the trouble (cc @superbobry).

**Reply by davisyoshida** [reply]
Great, thanks. I can take another shot at it once fix is in. Might not be until next weekend depending on how much free time I have this week.

**Reply by superbobry** [reply]
Sorry about that, Davis. The next nightly `jaxlib` should have the `triton.compat` subpackage included.

**Reply by davisyoshida** [reply]
I still haven't been successful in getting a version where I can run the tests, but now it's due to XLA not liking the current version of ptxas. In the event that I do get a working configuration at some point I'll still try to submit a PR, but I imagine it would be less effort for you guys to add the lowering yourselves than to help debug my config.

**Reply by sharadmv** [reply]
Sorry what do you mean by adding the lowering yourselves?

**Reply by davisyoshida** [reply]
I just meant adding the Triton lowering for bitcast_convert_type and the uint types. I don't know if it's worth you guys' time to unblock me on running the pallas tests on my setup.

**Reply by sharadmv** [reply]
I'm assuming if you can't run the tests, then you can't run other things though. 😔

**Reply by davisyoshida** [reply]
Well, the version set you posted the other day lets me use pallas, it's just not up to date.

**Reply by superbobry** [reply]
Can you share the error you are getting from XLA?

**Reply by davisyoshida** [reply]
I think that particular error might have been fixed when I upgraded my CUDA pip packages. I'm still having trouble getting a consistent version set though.

At least part of the problem seems to be that jax-triton needs an older version of triton that JAX is expecting at HEAD. For example, jax-triton tries to import `get_kernel_name` from `triton.compiler.backends.cuda`, which was removed openai/triton@8f64d4c28d7c5d036e4187bdb38df6219e3078de, but it also tries to use `triton._C.libtriton.ir.load_dialects`, which wasn't added until openai/triton@7d3f045045eb8d1a36e01d5b9ba26644304c02cf.
